File: PyEngineering/02_middleware_chain/02_middleware_chain_explanation.py
"""
02 · Middleware chain
======================

WHAT WE'RE BUILDING
--------------------
A small stack of composable **pure-ASGI middleware** (not `BaseHTTPMiddleware`
— see WHY below) wrapping a FastAPI app, in this order from outermost to
innermost:

    ExceptionHandlingMiddleware   # never let an unhandled exception reach the client
      -> RequestIDMiddleware      # tag every request with a correlation id
        -> AccessLogMiddleware    # structured JSON log line per request
          -> TimeoutMiddleware    # bound how long a handler may run
            -> AuthMiddleware     # reject unauthenticated requests
              -> app              # your route handlers

WHY PURE ASGI, NOT `BaseHTTPMiddleware`
-----------------------------------------
Starlette's `BaseHTTPMiddleware` is the "easy" way to write middleware (it
gives you a `request`/`call_next` shape that looks like Flask/Express), but
it has two real production costs worth knowing:

1. It buffers the *entire* response body in memory before your middleware's
   code can inspect it — this breaks streaming responses and wastes memory
   on large payloads.
2. Every layer of `BaseHTTPMiddleware` spawns its own `anyio` task group to
   bridge the callback-based `call_next` back into ASGI's message-passing
   model — measurable overhead per layer, per request.

Pure ASGI middleware (a callable class implementing
`__call__(self, scope, receive, send)`, wrapping an inner `app` of the same
shape) has none of this overhead and is what production middleware
(`GZipMiddleware`, `CORSMiddleware`, Sentry's ASGI integration, etc.) is
actually built from in Starlette itself. It's more code up front — you must
understand the ASGI message protocol (`http.request`, `http.response.start`,
`http.response.body`) — but it composes cleanly and doesn't buffer.

CONCEPTS COVERED
-----------------
- The ASGI 3-callable protocol (`scope`, `receive`, `send`) and how
  middleware wraps another ASGI app of the same shape
- `contextvars.ContextVar` for request-scoped correlation IDs that are safe
  under `asyncio` (unlike a plain module global, which would leak across
  concurrent requests sharing one event loop)
- Structured (JSON) logging via stdlib `logging` + a custom `Formatter`
- Exception-handling middleware as the *outermost* layer, so it can catch
  failures from every layer beneath it, including bugs in your own other
  middleware
- Enforcing a per-request timeout with `asyncio.wait_for` / `asyncio.timeout`
  and mapping cancellation to a clean `504`
- Auth middleware that inspects a header and short-circuits the chain by
  sending a response directly (never calling the inner `app`) on failure

THE ASGI MESSAGE PROTOCOL (what you're actually wrapping)
------------------------------------------------------------
For an HTTP request, the ASGI server calls your app as
`await app(scope, receive, send)`. `scope` is a dict describing the request
(method, path, headers, ...). `receive()` is an async callable yielding
request body chunks. `send(message)` is an async callable your app calls
(possibly multiple times) to emit the response:

    await send({"type": "http.response.start", "status": 200,
                 "headers": [(b"content-type", b"application/json")]})
    await send({"type": "http.response.body", "body": b'{"ok": true}'})

Middleware wraps this: it receives `(scope, receive, send)` from the layer
above, and calls `await self.app(scope, receive, wrapped_send)` where
`wrapped_send` is a function you write that intercepts/mutates outgoing
messages (e.g. to add a header) before forwarding them to the real `send`.
To short-circuit (never reach the inner app — e.g. a failed auth check),
just call `send(...)` directly yourself and return without calling
`self.app(...)` at all.

THE SPEC
--------
Implement five ASGI middleware classes plus a small demo FastAPI app to
exercise them:

    RequestIDMiddleware
        - Generate a `uuid.uuid4().hex` request id per HTTP request (skip
          non-"http" scope types, e.g. "lifespan", by passing through
          unchanged).
        - Store it in a `ContextVar` so route handlers / other middleware
          can read it via `request_id_var.get()`.
        - Add it as an `X-Request-ID` response header on the way out.
        - Reset the ContextVar (via the token from `.set()`) in a `finally`
          so it never leaks into an unrelated request reusing the same task.

    AccessLogMiddleware
        - Time the request (`time.monotonic()` before/after).
        - After the inner app completes (success OR exception — use
          try/finally), emit ONE structured JSON log line via a
          module-level `logging.Logger` containing at least: `request_id`,
          `method`, `path`, `status` (captured by intercepting
          `http.response.start`), and `duration_ms`.
        - If the inner app raised, `status` should reflect what actually
          got sent (if `ExceptionHandlingMiddleware` above this layer sent a
          500, that 500 is what gets logged) or `None`/`0` if nothing was
          sent before the exception propagated past this layer too.

    TimeoutMiddleware
        - Constructor takes `timeout_seconds: float`.
        - Wrap the call to the inner app in `asyncio.timeout(timeout_seconds)`
          (3.11+, preferred over `asyncio.wait_for` for this — no need to
          juggle a coroutine object) or `asyncio.wait_for`.
        - On `TimeoutError`, if no response has started yet (track this via
          your `send` wrapper), send a clean `504` JSON response. If a
          response already started, there's nothing safe left to send —
          just let the cancellation propagate (do not try to send a second
          `http.response.start`; ASGI servers will reject or ignore it).

    AuthMiddleware
        - Constructor takes a `set[str]` of valid bearer tokens.
        - Read the `Authorization` header from `scope["headers"]` (a list of
          `(bytes, bytes)` tuples — headers are ALWAYS lower-cased bytes in
          ASGI scope, unlike the mixed-case strings you see in an HTTP
          client's response object).
        - Expect the format `Bearer <token>`. Missing/malformed header or a
          token not in the valid set -> send `401` directly (never call the
          inner app) with body `{"detail": "unauthorized"}`.
        - A `/health` path (exact match) bypasses auth entirely — a common
          real-world carve-out for load-balancer health checks.

    ExceptionHandlingMiddleware
        - The OUTERMOST layer (wraps everything else, including the other
          four middleware, so a bug in any of them is also caught here).
        - Catch any `Exception` (not `BaseException` — let
          `asyncio.CancelledError`/`KeyboardInterrupt` propagate, they are
          not application errors) raised by the inner app.
        - Log it (with `exc_info=True` so the traceback is captured) via the
          same structured logger.
        - If no response has started yet, send a `500` JSON body
          `{"detail": "internal server error"}` — never leak the exception
          message or traceback to the client.
        - If a response already started, there's nothing safe to do but
          re-raise (same reasoning as TimeoutMiddleware).

    build_app() -> ASGIApp
        - Returns the WRAPPED ASGI callable, not the bare `FastAPI`
          instance — `app(scope, receive, send)` resolves via
          `type(app).__call__`, never an instance attribute, so there is
          no way to retrofit middleware onto an existing `FastAPI()`
          instance's call behavior after the fact. Build the routes on a
          `FastAPI()` instance, then wrap *that* and return the outermost
          wrapper.
        - A tiny demo app with:
            GET /health          -> {"status": "ok"}  (no auth required)
            GET /whoami          -> {"request_id": <the current request id>}
            GET /boom            -> raises RuntimeError("boom") on purpose
            GET /slow?seconds=N  -> `await asyncio.sleep(N)` then 200 (used
                                    to exercise the timeout middleware)
        - Wrap it with all five middleware in the order given at the top of
          this file. Remember: in ASGI, the FIRST middleware you wrap with
          becomes the OUTERMOST layer only if you wrap in the right order —
          work out (and comment on) which order of `app = X(app)` calls
          produces "ExceptionHandling is outermost."

ACCEPTANCE CRITERIA
--------------------
1. `GET /health` succeeds with no `Authorization` header (auth bypass).
2. Any other route without a valid `Authorization: Bearer <token>` header
   returns 401 and never reaches the route handler (verify via a route that
   would otherwise have an observable side effect, or simply that `/boom`
   with no auth returns 401, not 500).
3. `GET /whoami` (with valid auth) returns a `request_id` that also appears
   in the response's `X-Request-ID` header, and the two match.
4. `GET /boom` (with valid auth) returns 500 with a generic body — the
   `RuntimeError` message is never present in the response.
5. `GET /slow?seconds=<longer than the configured timeout>` (with valid
   auth) returns 504, not 200 and not a hung connection.
6. Every request produces exactly one structured log record with the
   correct `status` and a `duration_ms >= 0`.
7. Two concurrent requests never see each other's `request_id` (the
   ContextVar isolation actually works) — provable by firing concurrent
   requests and asserting each response's id is unique and self-consistent.
8. mypy-clean, black-formatted, ruff-clean.
"""

from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable, MutableMapping
from contextvars import ContextVar

from fastapi import FastAPI

# ASGI type aliases (the stdlib/Starlette don't export these as a stable
# public API at time of writing, so most ASGI middleware in the wild rolls
# its own aliases like this).
Scope = MutableMapping[str, object]
Message = MutableMapping[str, object]
Receive = Callable[[], Awaitable[Message]]
Send = Callable[[Message], Awaitable[None]]
ASGIApp = Callable[[Scope, Receive, Send], Awaitable[None]]

request_id_var: ContextVar[str | None] = ContextVar("request_id", default=None)

logger = logging.getLogger("middleware_chain")


class RequestIDMiddleware:
    """Tags every HTTP request with a per-request correlation id."""

    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # TODO: pass through unchanged for non-"http" scope types.
        # TODO: generate a uuid4 hex id, `request_id_var.set(...)`, and
        # `.reset(token)` in a finally block.
        # TODO: build a `send` wrapper that appends an `x-request-id` header
        # on the `http.response.start` message before forwarding it.
        raise NotImplementedError("TODO: implement RequestIDMiddleware")


class AccessLogMiddleware:
    """Emits one structured JSON log line per request."""

    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # TODO: pass through unchanged for non-"http" scope types.
        # TODO: time the call, capture the eventual status via a `send`
        # wrapper, and log exactly once in a finally block (covers both the
        # success and exception paths).
        raise NotImplementedError("TODO: implement AccessLogMiddleware")


class TimeoutMiddleware:
    """Bounds how long the inner app may take to produce a response."""

    def __init__(self, app: ASGIApp, *, timeout_seconds: float) -> None:
        self.app = app
        self.timeout_seconds = timeout_seconds

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # TODO: pass through unchanged for non-"http" scope types.
        # TODO: track whether a response has started (via a `send`
        # wrapper). On `TimeoutError` from `asyncio.timeout(...)`, send a
        # 504 JSON body IF nothing has started yet; otherwise re-raise.
        raise NotImplementedError("TODO: implement TimeoutMiddleware")


class AuthMiddleware:
    """Rejects requests without a valid bearer token (except /health)."""

    def __init__(self, app: ASGIApp, *, valid_tokens: set[str]) -> None:
        self.app = app
        self.valid_tokens = valid_tokens

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # TODO: pass through unchanged for non-"http" scope types and for
        # scope["path"] == "/health".
        # TODO: read scope["headers"] (list[tuple[bytes, bytes]], lower-cased
        # names), parse "authorization: Bearer <token>", and send a 401
        # directly (do not call self.app) if missing/invalid.
        raise NotImplementedError("TODO: implement AuthMiddleware")


class ExceptionHandlingMiddleware:
    """Outermost layer: converts any unhandled Exception into a clean 500."""

    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        # TODO: pass through unchanged for non-"http" scope types.
        # TODO: try/except Exception around `await self.app(...)`; track
        # whether a response has started via a `send` wrapper; on exception,
        # log with exc_info=True and send a 500 JSON body if nothing has
        # started yet, else re-raise.
        raise NotImplementedError("TODO: implement ExceptionHandlingMiddleware")


def build_app(*, valid_tokens: set[str], timeout_seconds: float = 2.0) -> ASGIApp:
    """Wire the demo FastAPI app with all five middleware, in the correct
    nesting order so ExceptionHandlingMiddleware ends up outermost."""
    app = FastAPI()

    # TODO: add the four demo routes (/health, /whoami, /boom, /slow).

    # TODO: wrap `app` with the five middleware. Note that FastAPI/Starlette
    # is itself a valid ASGI app, so `app` can be passed directly as the
    # `app` argument to your outermost middleware constructor — but the
    # ORDER you nest them in determines the runtime call order. Work out
    # (and comment) which nesting produces the order documented at the top
    # of this file. Return the final (outermost) wrapped callable, not
    # `app` itself.
    raise NotImplementedError("TODO: implement build_app")


# ---------------------------------------------------------------------------
# Hints
# -----
# - `scope["headers"]` entries are `(bytes, bytes)` with LOWER-CASED header
#   names — `b"authorization"`, never `b"Authorization"`.
# - You can send a complete response "by hand" from inside middleware with
#   two `send()` calls; write a small private helper
#   `_send_json(send, status, body)` and reuse it in Auth/Timeout/
#   ExceptionHandling middleware instead of repeating the two calls three
#   times.
# - `asyncio.timeout(seconds)` (3.11+) is a context manager — prefer it over
#   `asyncio.wait_for(coro, seconds)` when you're not holding a bare
#   coroutine object already (here you're awaiting `self.app(...)` inline,
#   which `asyncio.timeout` handles more naturally).
# - To know whether a response already started inside Timeout/Exception
#   middleware, set a `bool` flag (e.g. `started = False`) in an enclosing
#   scope and flip it to `True` inside your `send` wrapper on
#   `"http.response.start"`.
#
# Pitfalls
# --------
# - Using a plain module-level variable instead of `ContextVar` for the
#   request id — under `asyncio`, concurrent requests interleave on one
#   thread/event loop, and a plain global gets clobbered by whichever
#   request set it last, not the request currently running.
# - Forgetting `.reset(token)` on the ContextVar — without it, in contexts
#   where tasks are reused (some server/test-client setups), a stale
#   request id can leak into the next request.
# - Sending a second `http.response.start` after one has already gone out
#   (e.g. Timeout or ExceptionHandling firing after the inner app already
#   began streaming) — ASGI servers will raise or the message is simply
#   invalid. Always gate on the `started` flag.
# - Catching `BaseException` instead of `Exception` in
#   ExceptionHandlingMiddleware — this would swallow
#   `asyncio.CancelledError`, breaking cooperative cancellation (e.g. from
#   TimeoutMiddleware itself, or from the ASGI server shutting down).
# - Forgetting the `/health` bypass and health-checking infrastructure
#   (load balancers, k8s probes) start getting 401s and taking the service
#   out of rotation.
#
# Stretch goals
# -------------
# - Add a `RateLimitMiddleware` (foreshadows problem 15) between Auth and
#   the app, keyed by the bearer token.
# - Make `AccessLogMiddleware` also log request body SIZE (not content —
#   never log bodies verbatim in production, they may contain secrets/PII)
#   by summing `http.request` message body chunks as they pass through
#   `receive`.
# - Propagate `X-Request-ID` from an *inbound* header if the caller already
#   set one (common when this service is itself downstream of another
#   service in a call chain) instead of always generating a fresh one.

File: GoEngineering/02_middleware_chain/explanation/02_middleware_chain_explanation.go
/*
Problem 02 — Middleware Chain (log/slog, request IDs, recovery, timeouts, auth)

WHAT WE'RE BUILDING

A composable HTTP middleware stack for the kind of service problem 01 built:
structured request logging, per-request IDs threaded through context and
response headers, panic recovery that never lets a handler crash the
process, a request-scoped timeout, and a simple bearer-token auth gate —
all as independently testable `func(http.Handler) http.Handler` values
chained together in a defined order.

# WHY THIS MATTERS IN REAL SYSTEMS

Every production HTTP service needs the same handful of cross-cutting
concerns applied to (almost) every route: you don't want to copy-paste
"generate a request ID, log the request, recover from panics" into every
handler. Middleware is how Go solves this without a framework — plain
functions wrapping `http.Handler`, composed in a chain. Getting the *order*
right is what separates "logging shows real duration and status" from
"the logging middleware innocently reports 200 on responses that panicked
after it ran." This problem is also where structured logging
(`log/slog`) production patterns live: attaching request-scoped fields
(request ID, method, path, status, duration, remote addr) to every log
line rather than a bare `log.Println`.

CONCEPTS COVERED

  - Middleware type: `type Middleware func(http.Handler) http.Handler`
  - Chaining/composition: `Chain(mw1, mw2, mw3)(finalHandler)`
  - `context.Context` value propagation for a per-request ID
  - `log/slog` structured logging with request-scoped attributes
  - Panic recovery middleware (must run outermost-but-one — see spec)
  - Response status/byte-count capture via a wrapped `http.ResponseWriter`
    (the standard `http.ResponseWriter` doesn't expose what status code a
    handler wrote — you have to intercept `WriteHeader`)
  - Per-request timeout via `http.TimeoutHandler` vs a hand-rolled
    `context.WithTimeout` + goroutine (and why the two differ)
  - A trivial bearer-token auth middleware and 401 handling
  - `sync/atomic` for a lock-free request-ID counter

# SPEC

Build these middlewares, each `func(http.Handler) http.Handler`:

  - RequestID: generates (or forwards, if the client already sent
    `X-Request-ID`) a unique ID, stores it in the request context under an
    unexported key type, and sets it on the response header
    `X-Request-ID`.
  - Logger: logs one structured line per request (method, path, status,
    duration, request ID, remote addr) using `log/slog`, AFTER the handler
    returns — which means it needs to know the status code the handler
    wrote, hence a wrapping ResponseWriter.
  - Recoverer: recovers from any panic in a downstream handler, logs it
    (with a stack trace), and writes a 500 with the standard error
    envelope `{"error":{"code":"internal_error","message":"..."}}` instead
    of letting net/http's default panic handling close the connection
    with no response body written incorrectly (net/http already recovers
    panics per-request to avoid killing the whole server, but it just
    closes the connection with no JSON body and logs to stderr — this
    middleware gives you a controlled, on-brand response instead).
  - Timeout(d time.Duration): bounds how long a request may run; on
    timeout, responds 503 with the error envelope
    `{"error":{"code":"timeout","message":"request timed out"}}` and
    cancels the handler's context so downstream work (e.g. a slow DB call
    passed r.Context()) can observe cancellation and stop early.
  - Auth(token string): requires header `Authorization: Bearer <token>`
    matching exactly; on mismatch/missing, responds 401 with
    `{"error":{"code":"unauthorized","message":"..."}}` without calling
    the next handler.

Chain composition:
  - `Chain(mws ...Middleware) Middleware` — returns a single Middleware
    that applies each in order such that the FIRST middleware passed is
    the OUTERMOST (runs first on the way in, last on the way out). This
    is the same convention as chi/gorilla — most Go middleware chains
    read top-to-bottom as "outermost first."
  - Required outer-to-inner order for the demo server in main():
    RequestID -> Recoverer -> Logger -> Timeout -> (per-route Auth where
    needed) -> handler.
  - Order matters and is graded: RequestID must be outermost so the ID
    exists before Logger/Recoverer run (they log it); Recoverer must wrap
    Logger... actually think about this carefully: if Recoverer is
    OUTSIDE Logger, a panic never reaches Logger's deferred status-code
    read, so Logger would log nothing for panicked requests. Decide, and
    document in your solution, which of "Recoverer must see panics from
    Logger too" vs "every request (including panics) gets exactly one
    Logger line" you're optimizing for, and pick an order that achieves
    both (hint: Logger can defer-recover indirectly through the wrapped
    ResponseWriter status even if Recoverer runs inside it, since
    Recoverer writes the 500 response itself before returning normally).

ACCEPTANCE CRITERIA

  - Chain(a, b, c)(handler) executes a's pre-logic, then b's, then c's,
    then handler, then c's post-logic, then b's, then a's — verified by a
    test recording call order.
  - A handler that panics never crashes the test process and always
    produces a 500 with the exact error envelope.
  - Every response (including panicked and timed-out ones) has an
    `X-Request-ID` header.
  - Timeout actually cancels the handler's context (a handler that
    selects on `ctx.Done()` observes cancellation at the deadline, not
    just "the client got a response early while the handler kept running
    forever" — that's a goroutine leak).
  - Auth rejects missing/wrong tokens with 401 and allows the correct one
    through to the wrapped handler.
*/
package main

import (
	"context"
	"net/http"
	"time"
)

// Middleware wraps an http.Handler with additional behavior and returns a
// new http.Handler.
type Middleware func(http.Handler) http.Handler

// Chain composes middlewares so that mws[0] is outermost (runs first on
// the way in, last on the way out) and mws[len-1] is innermost (runs
// immediately before the final handler).
//
// TODO: implement by folding from the last middleware to the first, each
// time wrapping the accumulated handler.
func Chain(mws ...Middleware) Middleware {
	panic("TODO: implement Chain")
}

// ---------------------------------------------------------------------------
// Request ID
// ---------------------------------------------------------------------------

// TODO: define an unexported context key type (never use a bare string or
// other exported type as a context key — it collides across packages) and
// a package-level instance of it, e.g.:
//
//	type ctxKey int
//	const requestIDKey ctxKey = iota

// RequestIDHeader is the response/request header carrying the request ID.
const RequestIDHeader = "X-Request-ID"

// RequestID returns middleware that ensures every request has a unique ID:
// forwards an incoming X-Request-ID header if present, otherwise generates
// one. Stores it in the request context and sets it on the response.
//
// TODO: implement. Use a package-level atomic.Uint64 (or similar) for ID
// generation — no need for a real UUID library here.
func RequestID(next http.Handler) http.Handler {
	panic("TODO: implement RequestID middleware")
}

// RequestIDFromContext extracts the request ID stored by RequestID, or ""
// if none is present (e.g. called outside the middleware chain).
// TODO: implement using the unexported context key.
func RequestIDFromContext(ctx context.Context) string {
	panic("TODO: implement RequestIDFromContext")
}

// ---------------------------------------------------------------------------
// Response-status-capturing ResponseWriter
// ---------------------------------------------------------------------------

// statusRecorder wraps http.ResponseWriter to capture the status code and
// byte count written, since the standard interface exposes neither after
// the fact.
//
// TODO: embed http.ResponseWriter, add status/bytes fields, override
// WriteHeader to record status (and call through), and consider overriding
// Write too (to record bytes and to default the status to 200 if
// WriteHeader was never called explicitly — matching net/http's own
// behavior).
type statusRecorder struct {
	http.ResponseWriter
	// TODO: fields
}

func (r *statusRecorder) WriteHeader(code int) {
	panic("TODO: implement statusRecorder.WriteHeader")
}

func (r *statusRecorder) Write(b []byte) (int, error) {
	panic("TODO: implement statusRecorder.Write")
}

// ---------------------------------------------------------------------------
// Logger
// ---------------------------------------------------------------------------

// Logger returns middleware that logs one structured line per request via
// log/slog, including the status code and duration, after the handler
// returns.
// TODO: wrap w in a statusRecorder, call next, then log method/path/status/
// duration/request-id/remote-addr.
func Logger(next http.Handler) http.Handler {
	panic("TODO: implement Logger middleware")
}

// ---------------------------------------------------------------------------
// Recoverer
// ---------------------------------------------------------------------------

// Recoverer returns middleware that recovers panics from downstream
// handlers and responds 500 with the standard error envelope instead of
// letting the panic propagate (net/http's server would otherwise recover
// it silently per-connection and just close the connection with no body).
// TODO: implement with defer/recover. Log the recovered value and a stack
// trace (debug.Stack()).
func Recoverer(next http.Handler) http.Handler {
	panic("TODO: implement Recoverer middleware")
}

// ---------------------------------------------------------------------------
// Timeout
// ---------------------------------------------------------------------------

// Timeout returns middleware that bounds request handling to d, canceling
// the handler's context and responding 503 with the error envelope if the
// deadline is exceeded before the handler finishes.
//
// TODO: implement. You may use http.TimeoutHandler as a starting point but
// note it does NOT use the standard error envelope by default and its
// default body is a plain string — decide whether to use it with a custom
// message or hand-roll with context.WithTimeout + a done channel. Document
// your choice in the solution.
func Timeout(d time.Duration) Middleware {
	panic("TODO: implement Timeout middleware")
}

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

// Auth returns middleware that requires an exact bearer token match.
// TODO: implement. Compare using constant-time comparison
// (crypto/subtle.ConstantTimeCompare) to avoid a timing side-channel on
// token comparison — a real production detail, not paranoia, since this is
// literally the credential check.
func Auth(token string) Middleware {
	panic("TODO: implement Auth middleware")
}

// ---------------------------------------------------------------------------
// Demo server
// ---------------------------------------------------------------------------

func main() {
	// TODO:
	//  1. Build a mux with a couple of demo routes: GET /ping (returns 200
	//     immediately), GET /slow (sleeps longer than the configured
	//     timeout, to demonstrate Timeout firing), GET /boom (panics, to
	//     demonstrate Recoverer), GET /secret (wrapped in Auth).
	//  2. Chain(RequestID, Recoverer, Logger, Timeout(2*time.Second))
	//     around the mux; wrap /secret's handler individually with
	//     Auth("supersecret") before it reaches the shared chain, or after
	//     — decide and document which makes more sense (hint: Auth should
	//     run before you do expensive/slow work, so probably inside
	//     Timeout but the exact composition is your call).
	//  3. Serve with the same production-sane timeouts pattern as
	//     problem 01's newServer, with graceful shutdown on SIGINT/SIGTERM.
	panic("TODO: implement main")
}

/*
HINTS

  - Chain's fold direction is easy to get backwards. If mws = [A, B, C] and
    you want A outermost, build inside-out: h := final; h = C(h); h = B(h);
    h = A(h). That means you iterate mws in REVERSE when wrapping.
  - `context.WithValue` keys must never be a built-in type like string —
    two packages both using the key "requestID" would silently collide and
    overwrite each other's context values. Always use an unexported type
    defined in your own package.
  - `http.TimeoutHandler`'s handler runs in its own goroutine and there is
    no way to forcibly kill that goroutine if it ignores context
    cancellation — the same is true of any hand-rolled version. "Timeout"
    only means "stop waiting for the response," not "stop the goroutine."
    A well-behaved handler must itself select on ctx.Done() to actually
    stop working — document this limitation rather than pretend it away.
  - WriteHeader is only called explicitly by handlers that need a non-200
    status. If a handler just calls w.Write(...) without WriteHeader, the
    first Write implicitly sends a 200 — your statusRecorder must handle
    both cases to report status accurately.

COMMON PITFALLS

  - Calling WriteHeader more than once (e.g. Recoverer trying to write a
    500 after the panicking handler already wrote a 200 and some body
    bytes) — net/http logs "superfluous response.WriteHeader call" and the
    original bytes already went out. The fix is to only ever have ONE
    place in the chain capable of writing the final response for a given
    outcome, and to check whether headers were already sent before trying
    to override them (recoverer generally can't fix a response that
    already started streaming — document this limitation).
  - Recovering a panic but forgetting to also stop the goroutine/request
    from continuing further catastrophic work — recover() only stops the
    unwind at that stack frame; if you spawned other goroutines from the
    handler before panicking, they keep running and are now leaked/
    unsupervised.
  - Reading the request ID from context.Value with a string key instead of
    a typed unexported key — works until another package/middleware picks
    the same string, then values silently collide.
  - Putting Logger outside Recoverer — a panic'd request wrote no log line
    at all because Logger's own "log after next() returns" code path
    never executes if a panic unwinds straight through it. (This is
    exactly why the SPEC's required order nests Recoverer *inside*
    Logger — but Logger must still see the eventual status Recoverer
    wrote via the shared statusRecorder.)
  - Running the handler in a goroutine inside Timeout (needed to race it
    against the deadline) and forgetting that panics do NOT cross
    goroutine boundaries: an unrecovered panic in that spawned goroutine
    crashes the whole process even though Recoverer sits right there in
    the chain — Recoverer's recover() only catches panics unwinding its
    OWN goroutine's stack. You must recover inside the spawned goroutine
    yourself and re-panic on the original goroutine (the one Recoverer
    wraps) once you know the handler didn't time out, so the panic
    re-enters the normal chain. This is exactly the kind of bug that only
    shows up when you actually exercise a panicking handler through the
    full chain in a test — write that test.

STRETCH GOALS

  - Add a CORS middleware.
  - Add per-route rate limiting using golang.org/x/time/rate (already
    vendored in this module — see problem 15 for the general pattern).
  - Make Logger emit different slog levels based on status (5xx -> Error,
    4xx -> Warn, else Info).
  - Add OpenTelemetry-style trace/span ID propagation instead of a plain
    counter-based request ID.
*/
